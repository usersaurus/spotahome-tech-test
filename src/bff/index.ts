import axios from 'axios'

export interface IRawHouseData {
  Address: string
  City: string
  CountryCode: string
  Price: number
  StartDate: string
  EndDate: string
  Currency: string // should be 'USD' | 'EUR' | 'GBP' or something like that, but let's keep it simple
  Description: string
  Lat: number
  Lng: number
  Facilities: { [key: string]: 'yes' | 'no' }
  Deposit: number
  Images: string[]
  LastUpdated: string
  ListingReference: string
  Link: string
}

export const getHouses = () => {
  const url = 'http://feeds.spotahome.com/ads-housinganywhere.json'

  return axios.get<IRawHouseData[]>(url).then(({ data }) => data)
}
