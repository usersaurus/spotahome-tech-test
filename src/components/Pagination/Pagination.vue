<script setup lang="ts">
import { IPagination } from '../../bff'

interface IPaginationProps {
  paginationData: IPagination
}

const props = defineProps<IPaginationProps>()
const emit = defineEmits<{
  (e: 'pageChanged', page: number): void
}>()

const changePage = (page: number) => {
  if (page < 0 || page > props.paginationData.last) return

  emit('pageChanged', page)
}
</script>

<template>
  <div class="pagination">
    <div
      class="pagination--back"
      :class="{ disabled: paginationData.current === 1 }"
      @click="changePage(paginationData.current - 1)"
    >
      <i class="fa-solid fa-caret-left"></i>
    </div>
    <div class="pagination--current">{{ paginationData.current }}</div>
    <div
      class="pagination--next"
      :class="{ disabled: paginationData.current === paginationData.last }"
      @click="changePage(paginationData.current + 1)"
    >
      <i class="fa-solid fa-caret-right"></i>
    </div>
  </div>
</template>

<style lang="postcss">
.pagination {
  display: inline-flex;
  gap: 10px;
  justify-content: center;
  padding-top: 16px;
  width: 100%;
  font-weight: 600;
  font-size: 13pt;
}

.pagination--next,
.pagination--back {
  cursor: pointer;

  &.disabled {
    cursor: not-allowed;

    color: gray;
  }
}
</style>
