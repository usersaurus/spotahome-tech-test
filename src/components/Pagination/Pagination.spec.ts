import { mount } from '@vue/test-utils'
import { IPagination } from '../../bff'
import Pagination from './Pagination.vue'

const paginationData = {
  last: 100,
  current: 1,
  pageSize: 10,
  totalElements: 10000,
} as IPagination

const doubleRaf = (callback: FrameRequestCallback) => {
  requestAnimationFrame(() => {
    requestAnimationFrame(callback)
  })
}

describe('The Pagination component', () => {
  const wrapper = mount(Pagination, {
    props: {
      paginationData,
    },
  })

  it('should render correctly', () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  it('should have the prev button disabled when current page is 1', () => {
    expect(wrapper.find('.pagination--back').classes()).toContain('disabled')
  })

  it('should change accordingly when next page button is clicked', async () => {
    await wrapper.find('.pagination--next').trigger('click')

    doubleRaf(() => {
      expect(wrapper.find('.pagination--back').classes()).not.toContain(
        'disabled'
      )
      expect(wrapper.find('.pagination--current').text()).toBe('2')
      expect(wrapper.emitted()).toHaveProperty('pageChanged')
      expect(wrapper.emitted().pageChanged[0]).toEqual([2])
    })
  })
})
