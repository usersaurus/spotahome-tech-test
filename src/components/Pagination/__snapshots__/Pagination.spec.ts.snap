// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Pagination component should render correctly 1`] = `
<div
  class="pagination"
>
  <div
    class="pagination--back disabled"
  >
    <i
      class="fa-solid fa-caret-left"
    />
  </div>
  <div
    class="pagination--current"
  >
    1
  </div>
  <div
    class="pagination--next"
  >
    <i
      class="fa-solid fa-caret-right"
    />
  </div>
</div>
`;
